### PR TITLE
THRIFT-4618: Use poll() instead of select() in C++ TNonblockingServer if available

### PR DESCRIPTION
**Pull request by st0ke**
Got poll() code from b5ebcd1

### CODE REVIEW

**Comment by st0ke** [comment]
"continuous-integration/travis-ci/pr" failed not because of my changes. Please, help.

**Comment by jeking3** [comment]
Sometimes it needs a little help.  We can restart individual jobs.  I have done that.

**Comment by jeking3** [comment]
I don't think non-blocking servers are tested in the current cross test suite...

**Comment by st0ke** [comment]
Do I need to rebase the commit?

**Comment by jeking3** [comment]
Unlikely, there are some stubborn failures.   I will try to rekick.

**Comment by st0ke** [comment]
> I see changes for cmake - what about autoconf?

Done

**Comment by st0ke** [comment]
> I see changes for cmake - what about autoconf?

Done

**Comment by duynl58** [comment]
Select is slow.
Please consider epoll/kqueue instead.

**Comment by st0ke** [comment]
> Select is slow.
> Please consider epoll/kqueue instead.

1. Current realization use select(). I add poll() if available, because FD_SET is failed if descriptor is bigger than 1024.
2. Does epoll/kqueue do it faster than poll on only two descriptors? If so, does anybody need this small boost?

File: lib/cpp/src/thrift/server/TNonblockingServer.cpp [hunk 1]
```diff
@@ -1291,10 +1295,44 @@ bool TNonblockingIOThread::notify(TNonblockingServer::TConnection* conn) {
...
+    pfd.revents = 0;
+    ret = poll(&pfd, 1, -1);
+    if (ret < 0) {
+      return false;
+    } else if (ret == 0) {
+      continue;
+    }
+
+    if (pfd.revents & POLLHUP || pfd.revents & POLLERR) {
+      ::close(fd);
```
Comment (jeking3) [inline lib/cpp/src/thrift/server/TNonblockingServer.cpp]: Should this and the other instances of close instead be THRIFT_CLOSESOCKET like the other code path?

Comment (st0ke) [reply lib/cpp/src/thrift/server/TNonblockingServer.cpp]: Yes, it should. closesocket() is used on win platform.

**Comment by jeking3** [comment]
Sadly I think our coverage on non-blocking is lax, so I'm sure how well this is tested across platforms; and there's no cross test that runs on Windows.

**Comment by st0ke** [comment]
"Travis build" failed in D language tests. Do I need to do something with that?

**Pull request merged**